File: app/src/main/java/io/geeteshk/hyper/git/GitWrapper.java
/*
 * Copyright 2016 Geetesh Kalakoti <kalakotig@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.geeteshk.hyper.git;

import android.content.Context;
import android.support.design.widget.Snackbar;
import android.text.SpannableString;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.Status;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.diff.DiffFormatter;
import org.eclipse.jgit.lib.ObjectId;
import org.eclipse.jgit.lib.Ref;
import org.eclipse.jgit.lib.StoredConfig;
import org.eclipse.jgit.revwalk.RevCommit;
import org.eclipse.jgit.util.StringUtils;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import io.geeteshk.hyper.R;
import io.geeteshk.hyper.adapter.ProjectAdapter;

/**
 * Helper class to handle git functions
 */
public class GitWrapper {

    /**
     * Log TAG
     */
    private static final String TAG = GitWrapper.class.getSimpleName();

    /**
     * git init
     *
     * @param context context to make toast
     * @param repo repo to init
     */
    public static void init(Context context, File repo, View view) {
        try {
            Git git = Git.init()
                    .setDirectory(repo)
                    .call();
            Snackbar.make(view, context.getString(R.string.repo_init) + git.getRepository().getDirectory(), Snackbar.LENGTH_LONG).show();
        } catch (GitAPIException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
        }
    }

    /**
     * git add -A
     *
     * @param repo repo to stage files
     */
    public static void add(View view, File repo) {
        try {
            Git git = getGit(view, repo);
            if (git != null) {
                git.add()
                        .addFilepattern(".")
                        .call();
                Snackbar.make(view, R.string.added_to_stage, Snackbar.LENGTH_LONG).show();
            }
        } catch (GitAPIException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
        }
    }

    /**
     * git commit -m 'message'
     *
     * @param context context to make toast
     * @param repo repo to commit to
     * @param message git commit message
     */
    public static void commit(Context context, View view, File repo, String message) {
        new CommitTask(context, view, repo, new String[] {"Committing changes", "Committed successfully.", "Unable to commit files."}).execute(message);
    }

    private static String changeTextToNone(String text) {
        if (StringUtils.isEmptyOrNull(text)) {
            return "None\n";
        }

        return text;
    }

    /**
     * git status
     *
     * @param repo repo to view status of
     * @param t text views to set status to
     */
    public static void status(View view, File repo, TextView... t) {
        try {
            Git git = getGit(view, repo);
            if (git != null) {
                Status status = git.status()
                        .call();

                Set<String> conflicting = status.getConflicting();
                StringBuilder conflictingOut = new StringBuilder();
                for (String conflict : conflicting) {
                    conflictingOut.append(conflict).append("\n");
                }
                t[0].setText(changeTextToNone(conflictingOut.toString()));

                Set<String> added = status.getAdded();
                StringBuilder addedOut = new StringBuilder();
                for (String add : added) {
                    addedOut.append(add).append("\n");
                }
                t[1].setText(changeTextToNone(addedOut.toString()));

                Set<String> changed = status.getChanged();
                StringBuilder changedOut = new StringBuilder();
                for (String change : changed) {
                    changedOut.append(change).append("\n");
                }
                t[2].setText(changeTextToNone(changedOut.toString()));

                Set<String> missing = status.getMissing();
                StringBuilder missingOut = new StringBuilder();
                for (String miss : missing) {
                    missingOut.append(miss).append("\n");
                }
                t[3].setText(changeTextToNone(missingOut.toString()));

                Set<String> modified = status.getModified();
                StringBuilder modifiedOut = new StringBuilder();
                for (String mod : modified) {
                    modifiedOut.append(mod).append("\n");
                }
                t[4].setText(changeTextToNone(modifiedOut.toString()));

                Set<String> removed = status.getRemoved();
                StringBuilder removedOut = new StringBuilder();
                for (String remove : removed) {
                    removedOut.append(remove).append("\n");
                }
                t[5].setText(changeTextToNone(removedOut.toString()));

                Set<String> uncommitted = status.getUncommittedChanges();
                StringBuilder uncommittedOut = new StringBuilder();
                for (String uncom : uncommitted) {
                    uncommittedOut.append(uncom).append("\n");
                }
                t[6].setText(changeTextToNone(uncommittedOut.toString()));

                Set<String> untracked = status.getUntracked();
                StringBuilder untrackedOut = new StringBuilder();
                for (String untrack : untracked) {
                    untrackedOut.append(untrack).append("\n");
                }
                t[7].setText(changeTextToNone(untrackedOut.toString()));

                Set<String> untrackedFolders = status.getUntrackedFolders();
                StringBuilder untrackedFoldersOut = new StringBuilder();
                for (String untrackedf : untrackedFolders) {
                    untrackedFoldersOut.append(untrackedf).append("\n");
                }
                t[8].setText(changeTextToNone(untrackedFoldersOut.toString()));
            }
        } catch (GitAPIException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
        }
    }

    /**
     * git log
     *
     * @param repo to get commits from
     * @return list of commits
     */
    public static List<RevCommit> getCommits(View view, File repo) {
        Iterable<RevCommit> log = null;
        List<RevCommit> revCommits = new ArrayList<>();
        try {
            Git git = getGit(view, repo);
            if (git != null) {
                log = git.log()
                        .call();
            }
        } catch (GitAPIException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
            return null;
        }

        if (log != null) {
            for (RevCommit commit : log) {
                revCommits.add(commit);
            }
        }

        return revCommits;
    }

    /**
     * git branch
     *
     * @param repo repo to get branches from
     * @return list of branches
     */
    public static List<Ref> getBranches(View view, File repo) {
        List<Ref> branches = null;
        try {
            Git git = getGit(view, repo);
            if (git != null) {
                branches = git.branchList()
                        .call();
            }
        } catch (GitAPIException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
            return null;
        }

        return branches;
    }

    /**
     * git branch branchName
     *
     * @param context context to make toast
     * @param repo to create branch
     * @param branchName name of branch
     * @param checked switch to branch if it exists
     */
    public static void createBranch(Context context, View view, File repo, String branchName, boolean checked) {
        if (checked) {
            new CheckoutTask(context, view, repo, new String[] {"Creating new branch", "Checked out successfully.", "Unable to checkout."}).execute(String.valueOf(true), branchName);
        } else {
            try {
                Git git = getGit(view, repo);
                if (git != null) {
                    git.branchCreate()
                            .setName(branchName)
                            .call();
                }
            } catch (GitAPIException e) {
                Log.e(TAG, e.toString());
                Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
            }
        }
    }

    /**
     * git branch -d branches
     *
     * @param repo to delete branches from
     * @param branches to delete
     */
    public static void deleteBranch(View view, File repo, String... branches) {
        try {
            Git git = getGit(view, repo);
            if (git != null) {
                git.branchDelete()
                        .setBranchNames(branches)
                        .call();
            }
        } catch (GitAPIException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
        }
    }

    /**
     * git checkout branch
     *
     * @param context context to make toast
     * @param repo to checkout to branch
     * @param branch to checkout to
     */
    public static void checkout(Context context, View view, File repo, String branch) {
        new CheckoutTask(context, view, repo, new String[] {"Checking out", "Checked out successfully.", "Unable to checkout."}).execute(String.valueOf(false), branch);
    }

    public static String getCurrentBranch(View view, File repo) {
        String branch = "";
        try {
            Git git = getGit(view, repo);
            if (git != null) {
                branch = git.getRepository()
                        .getFullBranch();
            }
        } catch (IOException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
            return null;
        }

        return branch;
    }

    /**
     * git clone remoteUrl
     *
     * @param context context to make toast
     * @param repo to clone
     * @param adapter to refresh
     * @param remoteUrl to clone from
     */
    public static void clone(Context context, View view, File repo, ProjectAdapter adapter, String remoteUrl, String username, String password) {
        if (!repo.exists()) {
            new CloneTask(context, view, repo, adapter).execute(remoteUrl, username, password);
        } else {
            Snackbar.make(view, R.string.folder_exists, Snackbar.LENGTH_LONG).show();
        }
    }

    public static void push(Context context, View view, File repo, String remoteUrl, boolean[] options, String username, String password) {
        new PushTask(context, view, repo, new String[] {"Pushing changes", "Successfully pushed commits to remote.", "There was a problem while pushing commits."}, options).execute(remoteUrl, username, password);
    }

    public static void pull(Context context, View view, File repo, String remote, String username, String password) {
        new PullTask(context, view, repo, new String[] {"Pulling changes", "Successfully pulled commits from remote.", "There was a problem while pulling commits."}).execute(remote, username, password);
    }

    public static void fetch(Context context, View view, File repo, String remote, String username, String password) {
        new FetchTask(context, view, repo, new String[] {"Fetching remote " + remote, "Successfully fetched from " + remote + ".", "There was a problem while fetching from " + remote + "."}).execute(remote, username, password);
    }

    static Git getGit(View view, File repo) {
        try {
            return Git.open(repo);
        } catch (IOException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
        }

        return null;
    }

    private static StoredConfig getConfig(View view, File repo) {
        Git git = getGit(view, repo);
        if (git != null) {
            return git.getRepository().getConfig();
        }

        return null;
    }

    public static String getRemoteUrl(View view, File repo, String remote) {
        String url = "";
        StoredConfig config = getConfig(view, repo);
        if (config != null) {
            url = config.getString("remote", remote, "url");
        }

        return url;
    }

    public static ArrayList<String> getRemotes(View view, File repo) {
        ArrayList<String> remotes = null;
        StoredConfig config = getConfig(view, repo);
        if (config != null) {
            remotes = new ArrayList<>(config.getSubsections("remote"));
        }

        return remotes;
    }

    public static void addRemote(View view, File repo, String remote, String url) {
        StoredConfig config = getConfig(view, repo);
        if (config != null) {
            config.setString("remote", remote, "url", url);
            try {
                config.save();
            } catch (IOException e) {
                Log.e(TAG, e.toString());
                Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
            }
        }
    }

    public static void removeRemote(View view, File repo, String remote) {
        StoredConfig config = getConfig(view, repo);
        if (config != null) {
            config.unsetSection("remote", remote);
            try {
                config.save();
            } catch (IOException e) {
                Log.e(TAG, e.toString());
                Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
            }
        }
    }

    public static boolean canCommit(View view, File repo) {
        try {
            Git git = getGit(view, repo);
            if (git != null) {
                return git.getRepository().getRepositoryState().canCommit()
                        && git.status().call().hasUncommittedChanges();
            }
        } catch (GitAPIException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
        }

        return false;
    }

    public static boolean canCheckout(View view, File repo) {
        Git git = getGit(view, repo);
        return git != null && git.getRepository().getRepositoryState().canCheckout();
    }

    public static SpannableString diff(View view, File repo, ObjectId hash1, ObjectId hash2) {
        SpannableString string = null;
        Git git = getGit(view, repo);
        try {
            if (git != null) {
                OutputStream out = new ByteArrayOutputStream();
                DiffFormatter formatter = new DiffFormatter(out);
                formatter.setRepository(git.getRepository());
                formatter.format(hash1, hash2);
                string = new SpannableString(out.toString());
            }
        } catch (IOException e) {
            Log.e(TAG, e.toString());
            Snackbar.make(view, e.toString(), Snackbar.LENGTH_LONG).show();
        }

        return string;
    }
}

File: app/src/main/java/io/geeteshk/hyper/git/GitWrapper.kt
/*
 * Copyright 2016 Geetesh Kalakoti <kalakotig@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.geeteshk.hyper.git

import android.content.Context
import android.text.SpannableString
import android.view.View
import android.widget.TextView
import io.geeteshk.hyper.R
import io.geeteshk.hyper.extensions.snack
import io.geeteshk.hyper.ui.adapter.ProjectAdapter
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.api.errors.GitAPIException
import org.eclipse.jgit.diff.DiffFormatter
import org.eclipse.jgit.lib.ObjectId
import org.eclipse.jgit.lib.Ref
import org.eclipse.jgit.lib.StoredConfig
import org.eclipse.jgit.revwalk.RevCommit
import org.eclipse.jgit.util.StringUtils
import timber.log.Timber
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.lang.ref.WeakReference
import java.util.*

object GitWrapper {

    fun init(context: Context, repo: File, view: View) {
        try {
            val git = Git.init()
                    .setDirectory(repo)
                    .call()
            view.snack(context.getString(R.string.repo_init, git.repository.directory))
        } catch (e: GitAPIException) {
            Timber.e(e)
            view.snack(e.toString())
        }

    }

    fun add(view: View, repo: File) {
        try {
            val git = getGit(view, repo)
            git?.let {
                git.add()
                        .addFilepattern(".")
                        .call()
                view.snack(R.string.added_to_stage)
            }
        } catch (e: GitAPIException) {
            Timber.e(e)
            view.snack(e.toString())
        }

    }

    fun commit(context: Context, view: View, repo: File, message: String) {
        CommitTask(WeakReference(context), WeakReference(view), repo, arrayOf("Committing changes", "Committed successfully.", "Unable to commit files.")).execute(message)
    }

    private fun changeTextToNone(text: String): String {
        return if (StringUtils.isEmptyOrNull(text)) {
            "None\n"
        } else text

    }

    fun status(view: View, repo: File, vararg t: TextView) {
        try {
            val git = getGit(view, repo)
            git?.let {
                val status = git.status()
                        .call()

                val conflicting = status.conflicting
                val conflictingOut = StringBuilder()
                for (conflict in conflicting) {
                    conflictingOut.append(conflict).append("\n")
                }
                t[0].text = changeTextToNone(conflictingOut.toString())

                val added = status.added
                val addedOut = StringBuilder()
                for (add in added) {
                    addedOut.append(add).append("\n")
                }
                t[1].text = changeTextToNone(addedOut.toString())

                val changed = status.changed
                val changedOut = StringBuilder()
                for (change in changed) {
                    changedOut.append(change).append("\n")
                }
                t[2].text = changeTextToNone(changedOut.toString())

                val missing = status.missing
                val missingOut = StringBuilder()
                for (miss in missing) {
                    missingOut.append(miss).append("\n")
                }
                t[3].text = changeTextToNone(missingOut.toString())

                val modified = status.modified
                val modifiedOut = StringBuilder()
                for (mod in modified) {
                    modifiedOut.append(mod).append("\n")
                }
                t[4].text = changeTextToNone(modifiedOut.toString())

                val removed = status.removed
                val removedOut = StringBuilder()
                for (remove in removed) {
                    removedOut.append(remove).append("\n")
                }
                t[5].text = changeTextToNone(removedOut.toString())

                val uncommitted = status.uncommittedChanges
                val uncommittedOut = StringBuilder()
                for (uncom in uncommitted) {
                    uncommittedOut.append(uncom).append("\n")
                }
                t[6].text = changeTextToNone(uncommittedOut.toString())

                val untracked = status.untracked
                val untrackedOut = StringBuilder()
                for (untrack in untracked) {
                    untrackedOut.append(untrack).append("\n")
                }
                t[7].text = changeTextToNone(untrackedOut.toString())

                val untrackedFolders = status.untrackedFolders
                val untrackedFoldersOut = StringBuilder()
                for (untrackedf in untrackedFolders) {
                    untrackedFoldersOut.append(untrackedf).append("\n")
                }
                t[8].text = changeTextToNone(untrackedFoldersOut.toString())
            }
        } catch (e: GitAPIException) {
            Timber.e(e)
            view.snack(e.toString())
        }

    }

    fun getCommits(view: View, repo: File): List<RevCommit>? {
        var log: Iterable<RevCommit>? = null
        val revCommits = ArrayList<RevCommit>()
        try {
            val git = getGit(view, repo)
            git?.let {
                log = git.log()
                        .call()
            }
        } catch (e: GitAPIException) {
            Timber.e(e)
            view.snack(e.toString())
            return null
        }

        log?.let {
            revCommits += log!!
        }

        return revCommits
    }

    fun getBranches(view: View, repo: File): List<Ref>? {
        var branches: List<Ref>? = null
        try {
            val git = getGit(view, repo)
            git?.let {
                branches = git.branchList()
                        .call()
            }
        } catch (e: GitAPIException) {
            Timber.e(e)
            view.snack(e.toString())
            return null
        }

        return branches
    }

    fun createBranch(context: Context, view: View, repo: File, branchName: String, checked: Boolean) {
        if (checked) {
            CheckoutTask(WeakReference(context), WeakReference(view), repo, arrayOf("Creating new branch", "Checked out successfully.", "Unable to checkout.")).execute(true.toString(), branchName)
        } else {
            try {
                val git = getGit(view, repo)
                git?.branchCreate()?.setName(branchName)?.call()
            } catch (e: GitAPIException) {
                Timber.e(e)
                view.snack(e.toString())
            }

        }
    }

    fun deleteBranch(view: View, repo: File, vararg branches: String) {
        try {
            val git = getGit(view, repo)
            git?.branchDelete()?.setBranchNames(*branches)?.call()
        } catch (e: GitAPIException) {
            Timber.e(e)
            view.snack(e.toString())
        }

    }

    fun checkout(context: Context, view: View, repo: File, branch: String) {
        CheckoutTask(WeakReference(context), WeakReference(view), repo, arrayOf("Checking out", "Checked out successfully.", "Unable to checkout.")).execute(false.toString(), branch)
    }

    fun getCurrentBranch(view: View, repo: File): String? {
        var branch = ""
        try {
            val git = getGit(view, repo)
            git?.let {
                branch = git.repository.fullBranch
            }
        } catch (e: IOException) {
            Timber.e(e)
            view.snack(e.toString())
            return null
        }

        return branch
    }

    fun clone(context: Context, view: View, repo: File, adapter: ProjectAdapter, remoteUrl: String, username: String, password: String) {
        if (!repo.exists()) {
            CloneTask(WeakReference(context), WeakReference(view), repo, adapter).execute(remoteUrl, username, password)
        } else {
            view.snack(R.string.folder_exists)
        }
    }

    fun push(context: Context, view: View, repo: File, remoteUrl: String, options: BooleanArray, username: String, password: String) {
        PushTask(WeakReference(context), WeakReference(view), repo, arrayOf("Pushing changes", "Successfully pushed commits to remote.", "There was a problem while pushing commits."), options).execute(remoteUrl, username, password)
    }

    fun pull(context: Context, view: View, repo: File, remote: String, username: String, password: String) {
        PullTask(WeakReference(context), WeakReference(view), repo, arrayOf("Pulling changes", "Successfully pulled commits from remote.", "There was a problem while pulling commits.")).execute(remote, username, password)
    }

    fun fetch(context: Context, view: View, repo: File, remote: String, username: String, password: String) {
        FetchTask(WeakReference(context), WeakReference(view), repo, arrayOf("Fetching remote $remote", "Successfully fetched from $remote.", "There was a problem while fetching from $remote.")).execute(remote, username, password)
    }

    internal fun getGit(view: View, repo: File): Git? {
        try {
            return Git.open(repo)
        } catch (e: IOException) {
            Timber.e(e)
            view.snack(e.toString())
        }

        return null
    }

    private fun getConfig(view: View, repo: File): StoredConfig? {
        val git = getGit(view, repo)
        return git?.repository?.config

    }

    fun getRemoteUrl(view: View, repo: File, remote: String): String {
        var url = ""
        val config = getConfig(view, repo)
        config?.let {
            url = config.getString("remote", remote, "url")
        }

        return url
    }

    fun getRemotes(view: View, repo: File): ArrayList<String>? {
        var remotes: ArrayList<String>? = null
        val config = getConfig(view, repo)
        config?.let {
            remotes = ArrayList(config.getSubsections("remote"))
        }

        return remotes
    }

    fun addRemote(view: View, repo: File, remote: String, url: String) {
        val config = getConfig(view, repo)
        config?.let {
            config.setString("remote", remote, "url", url)
            try {
                config.save()
            } catch (e: IOException) {
                Timber.e(e)
                view.snack(e.toString())
            }

        }
    }

    fun removeRemote(view: View, repo: File, remote: String) {
        val config = getConfig(view, repo)
        config?.let {
            config.unsetSection("remote", remote)
            try {
                config.save()
            } catch (e: IOException) {
                Timber.e(e)
                view.snack(e.toString())
            }

        }
    }

    fun canCommit(view: View, repo: File): Boolean {
        try {
            val git = getGit(view, repo)
            git?.let {
                return git.repository.repositoryState.canCommit() && git.status().call().hasUncommittedChanges()
            }
        } catch (e: GitAPIException) {
            Timber.e(e)
            view.snack(e.toString())
        }

        return false
    }

    fun canCheckout(view: View, repo: File): Boolean {
        val git = getGit(view, repo)
        return git != null && git.repository.repositoryState.canCheckout()
    }

    fun diff(view: View, repo: File, hash1: ObjectId, hash2: ObjectId): SpannableString? {
        var string: SpannableString? = null
        val git = getGit(view, repo)
        try {
            git?.let {
                val out = ByteArrayOutputStream()
                val formatter = DiffFormatter(out)
                formatter.setRepository(git.repository)
                formatter.format(hash1, hash2)
                string = SpannableString(out.toString())
            }
        } catch (e: IOException) {
            Timber.e(e)
            view.snack(e.toString())
        }

        return string
    }
}
